File: backend/app/services/cache.py
"""A minimal in-process TTL cache for the read-side rollups in aggregate.py
and events.py's venue-name lookup.

Every `/admin/*` endpoint used to call `aggregate.load_all()` independently —
four full Firestore collection scans each — with no reuse even when two of
them land within the same request burst (the admin dashboard and
ManageRoles both fire several of these calls in parallel). This collapses
that into one real scan per TTL window, with explicit `invalidate()` calls
from every write path so admin screens still reflect a write immediately
rather than waiting out the TTL.

Deliberately not a real caching library — one process, one dict, `time.monotonic()`
for expiry. Redis or similar would be over-engineering for "a few hundred
registrations" (see CLAUDE.md); the TTL is a safety net, not the primary
correctness mechanism.
"""

import time
from typing import Any, Callable

_store: dict[str, tuple[float, Any]] = {}


def cached(key: str, ttl_seconds: float, loader: Callable[[], Any]) -> Any:
    """Return the cached value for `key` if it hasn't expired, else call
    `loader()`, store the result, and return it."""
    hit = _store.get(key)
    now = time.monotonic()
    if hit is not None and hit[0] > now:
        return hit[1]

    value = loader()
    _store[key] = (now + ttl_seconds, value)
    return value


def invalidate(key: str) -> None:
    """Drop a cached entry so the next read recomputes it. Safe to call even
    if nothing is cached under `key` yet."""
    _store.pop(key, None)

File: backend/services/cache.js
/** A minimal in-process TTL cache — same shape as services/cache.py, adapted
 * for async loaders (Firestore reads are async in Node, unlike the sync
 * Python client).
 *
 * One process, one Map, performance.now() for expiry. Not a real caching
 * library on purpose: Redis or similar would be over-engineering for "a few
 * hundred registrations" (see CLAUDE.md). The TTL is a safety net; the
 * primary correctness mechanism is the explicit invalidate() calls wired
 * into every write path.
 *
 * The cached entry holds a Promise, not just a resolved value — two calls
 * that land while the first fetch is still in flight share it instead of
 * both hitting Firestore, which the sync Python version didn't need to
 * worry about but a concurrent Node server does.
 */
const store = new Map();

export async function cached(key, ttlSeconds, loader) {
  const hit = store.get(key);
  const now = performance.now();
  if (hit && hit.expiresAt > now) {
    return hit.promise;
  }

  const promise = Promise.resolve().then(loader);
  store.set(key, { expiresAt: now + ttlSeconds * 1000, promise });
  try {
    return await promise;
  } catch (err) {
    // A failed load shouldn't poison the cache for the TTL window.
    store.delete(key);
    throw err;
  }
}

export function invalidate(key) {
  store.delete(key);
}
